keyboards/planck/keymaps/sigul/keymap.c

```diff
@@ -0,0 +1,384 @@
+    _______, _______, _______, _______, _______, _______, _______,    TG(_RAISE), _______, _______, _______,  _______
+),
+
+/* MOUSE
+ * ,-----------------------------------------------------------------------------------.
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |      |      |      |      | MOUSE|      |      |      |Scr Up|Scr Do|      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |But1  |      |      |But2  | Left | Down | Up   |Right |
+ * `-----------------------------------------------------------------------------------'
+ */
+
+[_MOUSE] = LAYOUT_ortho_4x12(
+    _______, _______, _______, _______, _______,    _______, _______, _______,    _______,    _______,       _______,     _______,
+    _______, _______, _______, _______, _______,    _______, _______, _______,    _______,    _______,       _______,     _______,
+    _______, _______, _______, _______, _______,    MOUSE,   _______, _______,    _______,    KC_MS_WH_DOWN, KC_MS_WH_UP, _______,
+    _______, _______, _______, _______, KC_MS_BTN1, _______, _______, KC_MS_BTN2, KC_MS_LEFT, KC_MS_DOWN,    KC_MS_UP,    KC_MS_RIGHT
+)
+
+};
+
+#ifdef AUDIO_ENABLE
+  float plover_song[][2]     = SONG(PLOVER_SOUND);
+  float plover_gb_song[][2]  = SONG(PLOVER_GOODBYE_SOUND);
+#endif
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+  return update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST);
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+  
+    case QWERTY:
+      if (record->event.pressed) {
+        print("mode just switched to qwerty and this is a huge string\n");
+        set_single_persistent_default_layer(_QWERTY);
+      }
+      return false;
+      break;
+    
+    case PHONE:
+      if (record->event.pressed) {
+          SEND_STRING("3931044785");
+      }
+      return false;
+      break;
+    case DESK:
+      if (record->event.pressed) {
+          SEND_STRING("desk" SS_LALT(";") "silviogulizia.com");
+      }
+      return false;
+      break;
+    case SGCOM:
+      if (record->event.pressed) {
+          SEND_STRING("https" SS_LSFT(".")SS_LSFT("7")SS_LSFT("7")"silviogulizia.com");
+      }
+      return false;
+      break;
+    case VIVERE:
+      if (record->event.pressed) {
+          SEND_STRING("https" SS_LSFT(".") SS_LSFT("7") SS_LSFT("7") "vivereintenzionalmente.com");
+      }
+      return false;
+      break;
+    
+    case SVIV:
+      if (record->event.pressed) {
+          SEND_STRING("silvio" SS_LALT(";") "vivereintenzionalmente.com");
+      }
+      return false;
+      break;
```

Comment:
I'm not sure how much of this you want in the public repo.  If it's easily accessible, that's one thing.  However,  I have a bit of code that you may be interested in: 

https://github.com/qmk/qmk_firmware/blob/master/users/drashna/readme_secrets.md

Reply:
really interesting, but requires a bit of work. I will work on it in the next few days, thanks for the tip

Reply:
@drashna any chance you can help with this? I found the instruction not so clear for a newbie. And especially: 

- should I replace KC_SECRETS_1 with my KC names?
- should I replace "test1" and "secret1" with something?
- should I replace the following code completely with mine or just the KC_SECRET_1 name?
`clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
        send_string_with_delay(secret[keycode - KC_SECRET_1], MACRO_TIMER);`

I am getting lots of errors. Any chance you could provide a simple example code or I can find somewhere?

Reply:
I need to update those docs then, for some of it. 

>should I replace KC_SECRETS_1 with my KC names?

Yes.  The names are just references for you.  They can be anything you want.  But in this case, they'd be your keycode names. 

> should I replace "test1" and "secret1" with something?

You can, but you could just leave it as "", I have it set to something so I know if it's working properly. 

>should I replace the following code completely with mine or just the KC_SECRET_1 name?

You should only need: 
```c
send_string_with_delay(secret[keycode - KC_SECRET_1], MACRO_TIMER);
```

Just return true, and that will handle the oneshot code

Reply:
@drashna I really appreciate it, but I am going slightly mad with this. I understand I first need to have a user space and create the secrets files inside this (adding them in the .gitignore with the path /users/myname: is this correct?). 

Still trying to make the user space working and I keep getting errors with tri layers state users declared two times, but can't find where.

If you have some time to help, I could help writing a guide for newbie. Ok, it's not such a great thing but I don't have anything else to offer. 

In particular, I don't understand how to add the code for my keycodes in the secrets.c. could you make an example? thanks in advance.

Reply:
@drashna I followed your instructions but was not able to make the code working with secrets. I pushed them to the repo just in case you could have a look to it. 

this is the error I got

`Compiling: users/sigul/secrets.c                                                                   users/sigul/secrets.c: In function 'process_record_secrets':
users/sigul/secrets.c:21:32: error: 'secret' undeclared (first use in this function); did you mean 'secrets'?
         send_string_with_delay(secret[keycode - QWERTY], MACRO_TIMER);
                                ^~~~~~
                                secrets
users/sigul/secrets.c:21:32: note: each undeclared identifier is reported only once for each function it appears in
users/sigul/secrets.c:21:58: error: 'MACRO_TIMER' undeclared (first use in this function); did you mean 'MACRO_NONE'?
         send_string_with_delay(secret[keycode - QWERTY], MACRO_TIMER);
                                                          ^~~~~~~~~~~
                                                          MACRO_NONE
users/sigul/secrets.c:26:5: error: case label not within a switch statement
     case PHONE: // Secrets!  Externally defined strings, not stored in repo
     ^~~~
users/sigul/secrets.c:31:7: error: break statement not within loop or switch
       break;
       ^~~~~
users/sigul/secrets.c: At top level:
users/sigul/secrets.c:33:5: error: expected identifier or '(' before 'case'
     case DESK: // Secrets!  Externally defined strings, not stored in repo
     ^~~~
users/sigul/secrets.c:37:7: error: expected identifier or '(' before 'return'
       return false;
       ^~~~~~
users/sigul/secrets.c:38:7: error: expected identifier or '(' before 'break'
       break;
       ^~~~~
users/sigul/secrets.c:39:3: error: expected identifier or '(' before '}' token
   }
   ^
users/sigul/secrets.c:40:5: error: expected identifier or '(' before 'case'
     case SGCOM: // Secrets!  Externally defined strings, not stored in repo
     ^~~~
users/sigul/secrets.c:44:7: error: expected identifier or '(' before 'return'
       return false;
       ^~~~~~
users/sigul/secrets.c:45:7: error: expected identifier or '(' before 'break'
       break;
       ^~~~~
users/sigul/secrets.c:46:3: error: expected identifier or '(' before '}' token
   }
   ^
users/sigul/secrets.c:47:5: error: expected identifier or '(' before 'case'
     case VIVERE: // Secrets!  Externally defined strings, not stored in repo
     ^~~~
users/sigul/secrets.c:51:7: error: expected identifier or '(' before 'return'
       return false;
       ^~~~~~
users/sigul/secrets.c:52:7: error: expected identifier or '(' before 'break'
       break;
       ^~~~~
users/sigul/secrets.c:53:3: error: expected identifier or '(' before '}' token
   }
   ^
users/sigul/secrets.c:54:5: error: expected identifier or '(' before 'case'
     case SVIV: // Secrets!  Externally defined strings, not stored in repo
     ^~~~
users/sigul/secrets.c:58:7: error: expected identifier or '(' before 'return'
       return false;
       ^~~~~~
users/sigul/secrets.c:59:7: error: expected identifier or '(' before 'break'
       break;
       ^~~~~
users/sigul/secrets.c:60:3: error: expected identifier or '(' before '}' token
   }
   ^
users/sigul/secrets.c:61:3: error: expected identifier or '(' before 'return'
   return true;
   ^~~~~~
users/sigul/secrets.c:62:1: error: expected identifier or '(' before '}' token
 }
 ^
In file included from users/sigul/secrets.c:4:0:
users/sigul/secrets.h:1:22: error: 'secrets' defined but not used [-Werror=unused-variable]
 static const char *  secrets[] = {
                      ^~~~~~~
cc1: all warnings being treated as errors
 [ERRORS]
 |
 |
 |
make[1]: *** [.build/obj_planck_rev6_sigul/secrets.o] Error 1
make: *** [planck/rev6:sigul] Error 1
Make finished with errors`

Reply:
done. in your docs you use secret instead of secrets in the function to call the secrets, and that drove me crazy trying to understand. hopefully I was helped on reddit by ridingqwerty